pyscf/solvent/pcm.py

```diff
@@ -0,0 +1,627 @@
+    dy = 30.0*x**2 - 60.0*x**3 + 30.0*x**4 
+    dy[x<0] = 0.0
+    dy[x>1] = 0.0
+    return dy
+
+def gradgrad_switch_h(x):
+    ''' 2nd derivative of h(x) '''
+    ddy = 60.0*x - 180.0*x**2 + 120*x**3
+    ddy[x<0] = 0.0
+    ddy[x>1] = 0.0
+    return ddy
+
+def gen_surface(mol, ng=302, vdw_scale=1.2):
+    '''J. Phys. Chem. A 1999, 103, 11060-11079'''
+    unit_sphere = numpy.empty((ng,4))
+    libdft.MakeAngularGrid(unit_sphere.ctypes.data_as(ctypes.c_void_p), ctypes.c_int(ng))
+
+    atom_coords = mol.atom_coords(unit='B')
+    charges = mol.atom_charges()
+    N_J = ng * numpy.ones(mol.natm)
+    R_J = numpy.asarray([vdw_scale*Bondi[chg] for chg in charges])
+    R_sw_J = R_J * (14.0 / N_J)**0.5
+    alpha_J = 1.0/2.0 + R_J/R_sw_J - ((R_J/R_sw_J)**2 - 1.0/28)**0.5
+    R_in_J = R_J - alpha_J * R_sw_J
+    
+    grid_coords = []
+    weights = []
+    charge_exp = []
+    switch_fun = []
+    R_vdw = []
+    norm_vec = []
+    area = []
+    gslice_by_atom = []
+    p0 = p1 = 0
+    for ia in range(mol.natm):
+        symb = mol.atom_symbol(ia)
+        chg = gto.charge(symb)
+        r_vdw = vdw_scale*Bondi[chg]
+        
+        atom_grid = r_vdw * unit_sphere[:,:3] + atom_coords[ia,:]
+        riJ = scipy.spatial.distance.cdist(atom_grid[:,:3], atom_coords)
+        diJ = (riJ - R_in_J) / R_sw_J
+        diJ[:,ia] = 1.0
+        diJ[diJ < 1e-8] = 0.0
+        fiJ = switch_h(diJ)
+        
+        w = unit_sphere[:,3] * 4.0 * PI
+        swf = numpy.prod(fiJ, axis=1) 
+        idx = w*swf > 1e-16
+
+        p0, p1 = p1, p1+sum(idx)
+        gslice_by_atom.append([p0,p1])
+        grid_coords.append(atom_grid[idx,:3])
+        weights.append(w[idx])
+        switch_fun.append(swf[idx])
+        norm_vec.append(unit_sphere[idx,:3])
+        xi = XI[ng] / (r_vdw * w[idx]**0.5)
+        charge_exp.append(xi)
+        R_vdw.append(numpy.ones(sum(idx)) * r_vdw)
+        area.append(w[idx]*r_vdw**2*swf[idx])
+    
+    grid_coords = numpy.vstack(grid_coords)
+    norm_vec = numpy.vstack(norm_vec)
+    weights = numpy.concatenate(weights)
+    charge_exp = numpy.concatenate(charge_exp)
+    switch_fun = numpy.concatenate(switch_fun)
+    area = numpy.concatenate(area)
+    R_vdw = numpy.concatenate(R_vdw)
+    
+    surface = {
+        'ng': ng,
+        'gslice_by_atom': gslice_by_atom,
+        'grid_coords': grid_coords,
+        'weights': weights,
+        'charge_exp': charge_exp,
+        'switch_fun': switch_fun,
+        'R_vdw': R_vdw,
+        'norm_vec': norm_vec,
+        'area': area,
+        'R_in_J': R_in_J,
+        'R_sw_J': R_sw_J,
+        'atom_coords': atom_coords
+    }
+    return surface
+
+def get_F_A(surface):
+    '''
+    generate F and A matrix in  J. Chem. Phys. 133, 244111 (2010)
+    '''
+    R_vdw = surface['R_vdw']
+    switch_fun = surface['switch_fun']
+    weights = surface['weights']
+    A = weights*R_vdw**2*switch_fun
+    return switch_fun, A
+
+def get_dF_dA(surface):
```

Comment:
Following gradients code are duplicated from the code in grad/pcm.py . Can they be removed?